### PR TITLE
Support duplicate imports by default in `wasm-tools component new`

### PR DESCRIPTION
**Pull request by alexcrichton**
This commit is a refinement to the support added in #2145 to support duplicate imports by default in wasm-tools, namely by avoiding changing any binary offsets in a module except for the import section. This implements a strategy of renaming imports in the module that goes into a component to be unique. Classification of what an import maps to is then ensured to use the original names, not the current names, in the module.

### CODE REVIEW

**Comment by alexcrichton** [comment]
cc @erikrose I had a long plane ride and figured I could give this a crack and I think it worked out pretty well but would be good to get your eyes on it too

File: crates/wit-component/src/encoding/dedupe.rs [hunk 1]
```diff
@@ -1,316 +1,166 @@
...
+        //
+        // If this isn't fixed by the time someone reads this and is angry about
+        // O(n^2), sorry.
+        //
+        // Otherwise the purpose of this loop is to determine a unique name for
+        // the new import, something that hasn't previously been seen before.
+        let mut new_name = import.name.to_string();
+        for i in 2.. {
+            new_name.truncate(import.name.len());
+            write!(new_name, " [v{i}]").unwrap();
```
[the remark this answers is not in the document]
Comment (alexcrichton) [reply crates/wit-component/src/encoding/dedupe.rs:135]: Oh? Where did you see a 256-length limit? Wasmparser should have a limit of 100,000 as opposed to 256 to match web implementations. (or so we thought matched...)

File: crates/wit-component/tests/components/deduplicate-imports/module.wat [hunk 2]
```diff
@@ -1,15 +1,37 @@
...
 (module
   ;; Import a wasi function twice.
-  (func $i (import "wasi-snapshot-preview1" "proc_exit") (param i32))
-  (func $j (import "wasi-snapshot-preview1" "proc_exit") (param i32))
+  (func $exit1 (import "wasi-snapshot-preview1" "proc_exit") (param i32))
+  (func $exit2 (import "wasi-snapshot-preview1" "proc_exit") (param i32))
+
+  ;; define `f2` before the second `f` import shows up
+  (import "cm32p2" "f" (func $f_v1))
+  (import "cm32p2" "f2" (func $f2 (result i32)))
```
[the remark this answers is not in the document]
Comment (alexcrichton) [reply crates/wit-component/tests/components/deduplicate-imports/module.wat:9]: Heh this test was actually originally for that where I appended "N" to the end of the import instead of ` [v2]`. The previous syntax, as seen here, can clash with actual names that show up in the component model. I then had the idea of using a syntax which is impossible to show up in the component model (e.g. a space and brackets) so there's actually no worry of collisions any more. I figured I'd leave the test though as it might as well serve for future-proofing purposes.

File: crates/wit-component/tests/components/deduplicate-imports/module.wat [hunk 3]
```diff
@@ -1,15 +1,37 @@
...
+  (import "cm32p2" "f" (func $f_v3))
+
+  ;; define two `g` imports before the "real" `g2` import shows up
+  (import "cm32p2" "g" (func $g_v1))
+  (import "cm32p2" "g" (func $g_v2))
+  (import "cm32p2" "g2" (func $g2 (result i32)))
 
   ;; Call both copies of the import. They should both end up pointed to the same one.
   (func
+    call $f_v1
```
[the remark this answers is not in the document]
Comment (alexcrichton) [reply crates/wit-component/tests/components/deduplicate-imports/module.wat:20]: Nah just wanted to make sure they were all referenced. In the original implementation of "just add N" to fully test the collision handling I needed to have different imports listed in different orders, but it's not really necessary any more but I figured I'd leave things.

File: crates/wit-component/src/encoding/dedupe.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (erikrose) [reply crates/wit-component/src/encoding/dedupe.rs:136]: Oops, excuse me: [the max is 2^32](https://www.w3.org/2021/11/wasm-stage/syntax/values.html#names). So, even more extreme a corner case.

File: crates/wit-component/tests/components/deduplicate-imports/module.wat (hunk 2 above)
[the remark this answers is not in the document]
Comment (erikrose) [reply crates/wit-component/tests/components/deduplicate-imports/module.wat:8]: I dig it.:-D Hadn't seen [the more limited import/export names](https://github.com/WebAssembly/component-model/blob/main/design/mvp/Explainer.md#import-and-export-definitions) for components before.

**Comment by erikrose** [comment]
Boy, I'd love to get this in. Is Right Now good for you? I juuuust submitted an internal PR to pin to that wit-component release, because we need its functionality for a goal we're chasing. If you merge this now and cut a release, I'll pin to that instead, and we can forget the CLI flag ever existed!

**Comment by alexcrichton** [comment]
Yeah I'm happy to merge whenever but repo settings require 1+ approval from someone with write access, and 3/4 maintainers were on vacation last week so it may take a moment to work through the backlog to get to this.

No urgency from my part though, I'm happy to just leave the argument there but hide it from `--help` and/or make it a noop. Just wanted to confirm whether such logic was necessary or not.

**Review by dicej — APPROVED** [review]

**Comment by alexcrichton** [comment]
@erikrose ok I've landed/published this as-is with dropping the CLI argument. If that causes issues though let me know and I can add it back in.

**Comment by erikrose** [comment]
I think it’s dandy. If you have a moment to cut a release (of wit-component at least, though I think it’s all tied together), I’ll skip straight to this approach and avoid any future concern about CLI compat.

**Comment by alexcrichton** [comment]
Oh [already done](https://crates.io/crates/wit-component)!
